Update to Blynk local server including config files

877dev opened the pull request:
The original version did not create any config files and did not support email.

This commit fixes that, by making use of  `directoryfix.sh` to create the config folder,
which contains two config files, `server.properties` and `mail.properties`.
Note that if they already exist they are not overwritten.

The Dockerfile has been updated to use the latest version 0.41.14,
and a parameter to use a custom `mail.properties`.

`server.properties` should work "as is" for most people, but can be changed as required.
`mail.properties` requires editing to add gmail credentials.
`service.yml` updated to map the configs to the volumes folder.

A new readme guide has been produced which needs adding to the main webpage, this can be found [here](https://gist.github.com/877dev/cfa52044336e1e44d74156ce8b1458aa#file-blynk-setup-guide-md)

Zephael86 reviewed (changes requested):
I am an inexperineced github and docker user. I applied your changes on my system and it seems fine.

Another question reqarding the Blynk Server version... what is the custom in github?
On my local NAS I had a script that checked always the latest available version on github.
(However version updates are not so frequent anymore).
Is it possible to add such a step here, or hardcoding the exact version is the preferred method?

877dev commented:
Hey there,

Glad it's working for you. I'm not sure is the answer, I'm also relatively new to this. 

Some releases are named 'latest' however this does not seem to be the case with Blynk. It's on my list to do to find this out,I believe I listed it in future improvements. 

If you find out be sure to let me know 👍

Zephael86 commented:
The two config files gets created with your changeset. After attaching into the container if I modify e.g.: the /config/mail.properties it immediately gets reflected on host's file system (/home/pi/IOTstack/volumes/blynk_server/data/config/mail.properties).

My problem is now that these changes are not reflected in blynk (0.41.14).
Whenever I open the server.properties or mail.properties on the admin ui, it shows default values. (Not the ones coming from the config files).
Changing something in mail.properties will net sync-ed into the config file.
Same applies to server.properties.

blynk gets started as you specified in the ENTRYPOINT, with specific data folder, server and mail config files.

877dev commented:
> The two config files gets created with your changeset. After attaching into the container if I modify e.g.: the /config/mail.properties it immediately gets reflected on host's file system (/home/pi/IOTstack/volumes/blynk_server/data/config/mail.properties).
> 
> 
> 
> My problem is now that these changes are not reflected in blynk (0.41.14).
> 
> Whenever I open the server.properties or mail.properties on the admin ui, it shows default values. (Not the ones coming from the config files).
> 
> Changing something in mail.properties will net sync-ed into the config file.
> 
> Same applies to server.properties.
> 
> 
> 
> blynk gets started as you specified in the ENTRYPOINT, with specific data folder, server and mail config files.
> 
> 

Hi,

I updated the guide but it's not been posted yet, find it here:

https://gist.github.com/877dev/cfa52044336e1e44d74156ce8b1458aa

Basically the Admin UI config files don't work, and there is no plan by the developers to fix it. Just ignore them. 

Any issues with my guide please let me know. 

Thanks

Zephael86 commented:
Hi,

wow. Nice.
Just enough details for whoever user Blynk before. Even Node-Red basics are there.

Under the developer section you mention 'blynk_create.sh'.
Can it be that it was the previous name of: '.templates/blynk_server/directoryfix.sh' ?

One idea came in my mind regarding future improvements.
How to migrate data from an existing Blynk local server into the container? I will do this in the upcoming days - maybe today. Shall I write down the steps?

877dev commented:
> Under the developer section you mention 'blynk_create.sh'.
> Can it be that it was the previous name of: '.templates/blynk_server/directoryfix.sh' ?

Yes that's correct, I changed it now thank you. I realised there was a standard way of creating templates, and changed it to suit.



> How to migrate data from an existing Blynk local server into the container? I will do this in the upcoming days - maybe today. Shall I write down the steps?

Yes that sounds interesting, I can add the steps if you are happy it works. I don't have an existing "non-docker" local server to test it on.

Zephael86 commented:
Ok, I am going to create a guide how to migrate existing data (if local data and not postgres db is used).
Yesterday I was able to move my stuff from my NAS to the Raspberry PI.
After changing port forward rules on the router both the deployed uControllers and the Android Blynk App could connect without any change.

Back to the config files, and the editor...
I can confirm that your changes are working. Both the server.properties and the mail.proerties are parsed by Blynk.
First one is also displayed in the editor. And obviously, if the existing uCs can connect, then it is fine.
However the mail.properties editor displays some default values, instead of the content of the file. Very misleading.
The real settings were understood by Blynk, because I was able to send mail to my gmail account.

In the past I set the content of these files via text editor on my NAS. I was not even aware that the UI editor does not work.
Went back to my NAS, and checked the content of the mail.properties on the UI.
There, I could see the real content of the file. Same Blynk version.

Difference is the startup method compared to the container.
On the NAS the two files are besides the server.jar in the same folder.

After changing the Docker file a bit the mail.properties editor displayed the real content of the file.

FROM adoptopenjdk/openjdk11
MAINTAINER 877dev <877dev@gmail.com>`
ENV BLYNK_SERVER_VERSION 0.41.14
RUN mkdir /blynk
RUN curl -L https://github.com/blynkkk/blynk-server/releases/download/v${BLYNK_SERVER_VERSION}/server-${BLYNK_SERVER_VERSION}.jar > /blynk/server.jar
RUN mkdir /data
RUN mkdir /config && touch /config/server.properties
VOLUME ["/config", "/data/backup"]
**RUN ln -s /config/mail.properties /blynk/mail.properties && ln -s /config/server.properties /blynk/server.properties**
EXPOSE 8080 9443
WORKDIR /data
**ENTRYPOINT ["java", "-jar", "/blynk/server.jar", "-dataFolder", "/data"]**

Zephael86 commented:
Regarding

> Update: I added an [issue ](https://github.com/blynkkk/blynk-server/issues/1377) to blynk github.

Something like this works for me since 2016 on my NAS. Please don't hit me, I am not proud of it...
awk part can be removed if '.jar' gets removed from second curl call.

BLYNK_SERVER_VERSION=`curl -s https://github.com/blynkkk/blynk-server/releases |  grep server | grep -v client  | grep '.jar' | grep -v 'java8' | head -1 | cut -d'/' -f7- |  cut -d'"' -f1 | cut -d'-' -f2- | awk 'BEGIN {FS=".jar";}{print $1}'` && curl -L https://github.com/blynkkk/blynk-server/releases/download/v${BLYNK_SERVER_VERSION}/server-${BLYNK_SERVER_VERSION}.jar > /blynk/server.jar

Slyke commented on .templates/blynk_server/Dockerfile:
```diff
@@ -1,11 +1,11 @@
 FROM adoptopenjdk/openjdk11
-MAINTAINER Graham Garner <garner.gc@gmail.com>
+MAINTAINER 877dev <877dev@gmail.com>
```
Do you really want to become the maintainer? I should create an iotstack email, or have Andreas create one.

877dev replied:
@Slyke not not really :) I just put that there to differentiate my differences from the original. Feel free to change whatever you need to including port number quotations, I am just happy to see it getting updated and help a little!

Zephael86 replied:
Can you please consider adding symlinks inside the /blynk folder and changing the ENTRYPOINT as shown above?
Result would be that the mail.properties file content will be displayed in the editor.

Zephael86 replied:
Please change base image to 'adoptopenjdk/openjdk14'. According to #162

Slyke commented on .templates/blynk_server/directoryfix.sh:
```diff
@@ -0,0 +1,81 @@
+    #Give permissions:
+    sudo chown -R $u:$u ./volumes/blynk_server/data/config
+
+    #Populate the server.properties file:
```
While you can use double quote for multiline each, it might be better to use this method instead, incase a double quote is needed somewhere in the file:
```
#Populate the server.properties file:
cat <<EOT > ./volumes/blynk_server/data/config/server.properties
    hardware.mqtt.port=8440
    http.port=8080
    force.port.80.for.csv=false
    force.port.80.for.redirect=true
    https.port=9443
    data.folder=./data
    logs.folder=./logs
    log.level=info
    user.devices.limit=10
    user.tags.limit=100
    user.dashboard.max.limit=100
    user.widget.max.size.limit=20
    user.message.quota.limit=100
    notifications.queue.limit=2000
    blocking.processor.thread.pool.limit=6
    notifications.frequency.user.quota.limit=5
    webhooks.frequency.user.quota.limit=1000
    webhooks.response.size.limit=96
    user.profile.max.size=128
    terminal.strings.pool.size=25
    map.strings.pool.size=25
    lcd.strings.pool.size=6
    table.rows.pool.size=100
    profile.save.worker.period=60000
    stats.print.worker.period=60000
    web.request.max.size=524288
    csv.export.data.points.max=43200
    hard.socket.idle.timeout=10
    enable.db=false
    enable.raw.db.data.store=false
    async.logger.ring.buffer.size=2048
    allow.reading.widget.without.active.app=false
    allow.store.ip=true
    initial.energy=1000000
    admin.rootPath=/admin
    restore.host=blynk-cloud.com
    product.name=Blynk
    admin.email=admin@blynk.cc
    admin.pass=admin
EOT
```

Slyke commented on .templates/blynk_server/service.yml:
```diff
@@ -3,8 +3,9 @@
-      - "8180:8080"
-      - "8441:8441"
-      - "9443:9443"
+      - 8180:8080
```
Ports must have double quotes around them.

Zephael86 replied:
Really? I am not sure about this.
In the examples what I have seen there were no double quotes.

877dev replied:
Yup I realised this after submitting the PR, Grahams ogirinals did not have quotes but I noticed in the code changes you had updated for quotes".

Paraphraser commented:
Please see [documentation](https://docs.docker.com/compose/compose-file/#ports). In some situations unquoted port numbers can be misinterpreted. Rather than expecting everyone to remember two rules, the Docker people recommend always quoting. We have accepted that recommendation for IOTstack.

There may be some examples in .templates that are unquoted but those are the exception rather than the rule.

Zephael86 commented:
> Please see [documentation](https://docs.docker.com/compose/compose-file/#ports). In some situations unquoted port numbers can be misinterpreted. 

Thank you! Was looking for this answer!

Zephael86 commented:
@877dev I have posted some migration related steps here: [link](https://gist.github.com/877dev/cfa52044336e1e44d74156ce8b1458aa#file-blynk-setup-guide-md)

Zephael86 commented:
Hi All,
Can you please tell me how are pull merges handled? Is there any regular basis when these change requests are review-ed and merged into main?

liamhogan commented:
Hi all,
I'm new to docker (and github) and have been trying to get blynk_server working on IOTstack with no luck (see #162 ). 
It seems like your pull request @877dev  addresses some issues that might help me. 
Could someone give me some advice on how I can implement these changes on my system to try it out? I'm assuming there's a better way than manually creating the scripts and dockerfile and copying over the text from your commit!

Paraphraser commented:
@liamhogan 

> Could someone give me some advice on how I can implement these changes on my system

This is how I do it. First, go to the pull request page you're interested in and figure out its ID. In this case:

> Update to Blynk local server including config files #131

so we're talking about pull request 131. See how I've substituted that into the following:

```
$ cd ~/IOTstack
$ git checkout master
$ git pull origin master
$ git fetch origin pull/131/head:PR131
$ git checkout PR131
```

The `/131/` is the ID while `PR131` is just a branch name and could be anything but I always do it like this because I like to keep things simple and predictable.

When you want to revert (eg when the PR makes it into the master):

```
$ git checkout master
$ git branch -D PR131
```

<hr>

If you want to do this with multiple PRs, you do as many `git fetch` calls as are needed but don't do the `git checkout` of each PR. Instead, create a "holding" branch:

```
$ git checkout -b merge-pull-requests 
```

and then merge as many PRxxx branches as you need:

```
$ git merge -m "local merge branch" PR131
...
```

If you get a conflict then this command is your friend:

```
$ git merge --abort
```

Merge conflicts can sometimes be resolved by changing the order in which you do the `git merge` but it's also sometimes useful to use one of the following forms:

```
$ git merge -X theirs -m "local merge branch" PR131
$ git merge -X   ours -m "local merge branch" PR131
```

depending on whether you want the PR being merged to override, or be overridden by, what's already there.

When you're done, cleanup:

```
$ git checkout master
$ git branch -D merge-pull-requests
$ git branch -D PR131
...
```

If anyone knows of any smarter ways of doing these things, please let me know.
